### ISSUE TITLE
Bidirectional DC lines

### ISSUE THREAD

**Issue by fbonhoff** [e2nIEE/pandapower#2716]
### Bug report checklis

- [x] Searched the [issues page](https://github.com/e2nIEE/pandapower/issues) for similar reports

- [x] Read the relevant sections of the [documentation](https://pandapower.readthedocs.io/en/latest/about.html)

- [x] Browse the [tutorials](https://github.com/e2nIEE/pandapower/tree/develop/tutorials) and [tests](https://github.com/e2nIEE/pandapower/tree/develop/pandapower/test) for usefull code snippets and examples of use

- [x] Reproduced the issue after updating with `pip install --upgrade pandapower` (or `git pull`)

- [x] Tried basic troubleshooting (if a bug/error) like restarting the interpreter and checking the pythonpath

### Reproducible Example

```python
import pandapower as pp
net = pp.networks.case9()
pp.create_dcline(
                net,
                from_bus=4,
                to_bus=8,
                p_mw=-5,
                loss_percent=10,
                loss_mw=0.0,
                vm_from_pu=1,
                vm_to_pu=1,
                max_p_mw=10,
                min_q_to_mvar=0,
                min_q_from_mvar=0,
                max_q_to_mvar=0,
                max_q_from_mvar=0,
            )

pp.runpp(net, 
        tolerance_mva=1e-8, 
        enforce_q_lims=True, 
        distributed_slack=False, 
        algorithm="nr", 
        max_iteration=10, 
        init="auto",
        )

print(sum(abs(net.res_line["p_from_mw"])) + sum(abs(net.res_dcline["p_from_mw"]))  + sum(abs(net.res_trafo["p_hv_mw"])))

print("----------------------------------------------------------------------")

net = pp.networks.case9()
p_in = 5
loss_percent=10
p_out = 5/(1-loss_percent/100)
pmax = 5
pp.create_gen(net, bus=8, p_mw=-p_out, vm_pu=1,
            min_p_mw=0, max_p_mw=pmax,
            max_q_mvar=0, min_q_mvar=0,
            in_service=True)

pp.create_gen(net, bus=4, p_mw=p_in, vm_pu=1,
            min_p_mw=-pmax, max_p_mw=0,
            max_q_mvar=0, min_q_mvar=0,
            in_service=True)

pp.runpp(net, 
        tolerance_mva=1e-8, 
        enforce_q_lims=True, 
        distributed_slack=False, 
        algorithm="nr", 
        max_iteration=10, 
        init="auto",
        )

print(sum(abs(net.res_line["p_from_mw"])) + sum(abs(net.res_dcline["p_from_mw"]))  + sum(abs(net.res_trafo["p_hv_mw"])))
```

### Issue Description and Traceback

So both versions of the network run but lead to different results.  This is because eventhoug accepted the dc_line does not handle negative values in a reasonable way. 

To be more precise, runpp first checks for dc_lines and than "_add_auxiliary_elements(net)" which calls  "_add_dcline_gens(net)" leading to the addition of two generator on each end of the dc line. 
But if a negative value for p_mw is given, the loss is still subtracted leading to less power being drawn on the to bus than received at the from bus, effectively generating power over the line. 

It is quite an easy fix of course, but what is interessting is, the expected behavioiur.

### Expected Behavior

In my understanding it makes sense to always consider the DC line from the perspective of the From Bus, so that this should be done:

        pfrom = -Pf
        if Pf<0
            # No check for maximum implemented yet. 
            # Should be like this: if -Pf > Pmax: throw error
            pto = (Pf-Loss1)/(1-Loss1/100)
        else
            pto = (Pf * (1-Loss1/100)) - Loss0
        end

So that if -5 is given, the generator at the from bus receives 5 MW and the to bus "sends" 5.55 (for 10 percent loss and no constant loss). 

This would lead to a Bidirectional interpretation of the line.

Alternatively it would be reasonable to change to and from bus whenever p_mw is negative. and keep the logic outherwise the same. 
To me this would be less intuitive, as it changes the interpretation of the line and the beheaviour around 0MW is not as smooth but has a jump. 

Third option that i can think of is to just send an error and let everybody handle the necessary conversion themselves. 

I personally prefer the first option, the only thing that might be troublesome is that the calculated power of the generator is larger than the input, so that if sweeps or optimizations are done and not carefully considered error appear but i guess you need to let the user do some considerations ... 

I have not come up with a good plan to incorporate all and imho it is a matter of preference how dc lines should be handled but I am looking forward to hear other perspectives

### Installed Versions

Python 3.10.12
Pandas 2.2.3
networkx 3.4.2
Scipy 1.13.1
NumPy 1.26.4
packaging 24.2
tqdm 4.67.1
deepdiff 8.4.2
Pandapower 3.1.2

### Label

- [ ] Relevant labels are selected

**Comment by KS-HTK** [comment]
@vogt31337 I remember you had some Issues with DC lines "generating" power, might this be related?

**Comment by vogt31337** [comment]
@KS-HTK, probably. 

@fbonhoff, thanks for bringing this up. I think that is something we need to fix. Until now I would just "reverse" the line, but this not really practical, especially if the flow reverses during a timeseries simulation.

**Comment by fbonhoff** [comment]
timeseries is a relevant usecase. I have added a fix in my branch. 
Which one would you call reverse the line? 
 1. Change "from" and "to" line
 2. Calculate loss reversevly 

I have the second version currently in my dev branch.

**Comment by fbonhoff** [comment]
Maybe this underlines what i mean, or makes it clearer. In the first case the power added to the hvdc line would correlate with those power at the from bus:

1.:  
With a constant loss of 0 MW and a percentage loss of 10 %:

<img width="623" height="496" alt="Image" src="https://github.com/user-attachments/assets/55acdc9a-298e-4201-a851-64d8f851f48e" />

<img width="729" height="571" alt="Image" src="https://github.com/user-attachments/assets/bfbf9cd0-5d7d-4cf3-8c86-65dde20a2001" />

With a constant loss of 2 MW and a percentage loss of 10 %:

<img width="623" height="494" alt="Image" src="https://github.com/user-attachments/assets/c3bafd08-3259-4648-a23b-8b8329e59e72" />

<img width="623" height="488" alt="Image" src="https://github.com/user-attachments/assets/48dd5e4a-e4d2-4b01-a878-15d4203a8ced" />


In the second case with the reverse calculation of the loss, so the second case this would look like this:
2.: 
With a constant loss of 0 MW and a percentage loss of 10 %:

<img width="702" height="540" alt="Image" src="https://github.com/user-attachments/assets/c7f0d46d-ee20-4ae9-99d9-9bae5eefef67" />

<img width="702" height="540" alt="Image" src="https://github.com/user-attachments/assets/e035db25-8673-4a1f-ba9b-3a94af5498f6" />

With a constant loss of 2 MW and a percentage loss of 10 %:

<img width="702" height="541" alt="Image" src="https://github.com/user-attachments/assets/7560c73f-757d-4793-92ce-93ad93206769" />

<img width="702" height="540" alt="Image" src="https://github.com/user-attachments/assets/ac3c764a-a4e8-4d08-960a-6d95e1cbc918" />

**Comment by vogt31337** [comment]
@fbonhoff , I looked at your PR #2731, I think it will be complicated when running a timeseries simulation. You've added a mode parameter, which has a lot of changes in the network_structure, datamodel and convert / load functions as a consequence. Therefore I would try to solve this differently.

Next thing: I think when you a constant loss of 2 MW, there will be a plateu of 2MW, and a power less than 2MW will be "eaten up" by the loss. What do you think?

**Comment by fbonhoff** [comment]
Mhh, why would that cause an issue? Shouldnt it just be set once and than the line would beheave in the chosen way? 
But yeah i was trying to quickly get through all the elements that needed to be changed but it was the only non determinate way i came up with fixing it statisfying "all" scenarios... 

yeah, thats right, thats also not handeled in the current implementation is it? 
The graphs should nevertheless illustrate the general ideology, i should have made a grid to clarify the comparison though...

I think it should in the end be possible to:
1. set any value below +- pmax of the line
2. be able to choose the beheaviour if p = p_from or p=p_line
3. if p < p_loss with p_loss = p_loss1 + p_loss2 * p , that energy is solely taken out of the network

That sums up the current "problems" with the dc_lines found through this issue i think. 
Unfortunately I am not sure how that would effect the control aspect of e.g. vsc lines

**Comment by vogt31337** [comment]
@fbonhoff, I think I still do not really get your problem... So on the one hand the documentation clearly states p_mw > 0.
So your case is basically not covered, and considered a "wrong use". But I agree it would be good to have the possibility to reverse the flow, without recreating the dcline.

For your last question, vsc lines (I am not sure what you mean by this) are not modeled with DC Line, they use line_dc (confusing I know...), which is basically a multi terminal HVDC system.

For your points:
1. Afaik a line does not have a pmax, but a maximum current. This DC Line approach does not really model a line, therefore this value does not exist.
2. what do you mean by p_from and p_line? Is this the reference to be taken?
3. In this case, I think 0 power will be transmitted (I do not know what p_loss1 and p_loss2 reference to). Therefore 0 power will be on both sides of the station. We could discuss and change the documentation so that loss_mw reflects the converter losses. They will still draw current even if nothing is transported. Also reactive power is not really modeled in this approach, I am no expert in HVDC interconnections, but I think this can get serious at some point.

I would adapt the code from PR #2731 to be able to calculate a powerflow in both directions. In a timeseries calculation I could imagine a controller changing that value constantly and also reversing the flow.

Maybe you could create a test which shows this behaviour and add it to that PR.

**Comment by fbonhoff** [comment]
I actually didnt know that about the documentation. But I agree that it should be handled someway.

Ok I thought they would be based on the same structure, I just wanted to emphasize that I am not totally informed on all aspects and how some changes might effect things down the line.

1. Yeah but generators do and its considered when creating an HVDC line: 
`for dctab in net.dcline.itertuples():
...
pmax = dctab.max_p_mw`

And it also part of the parameter set for 'create_dcline'. 

2. Mhh, i am lacking the words to really explain but let me try with an example: 
If we take the current implementationa nd create a dc line we set a parameter called p_mw, which will be taken as the power in MW for the generator attached to the from_bus. Now we could consider p_mw to always be that, so that if it is negative, we still attach a generator drawing power from the network of the from_bus. 
Alternatively we could consider p_mw to be the powerflow into the DC line, so that is not connected to the from bus anymore but to 'to the line'. This has the same result for positive p_mw's but for negative p_mw that would lead to the abs(p_mw) being drawn from the to_bus now and the from bus receiving ( abs(p_mw) - p_loss). 
So I tried to express the two optione as p_mw = p_from and p_mw = p_line respectively. I still think both have their relevance which I tried to express through the additional parameter, which causes the problems you described. 
My first impression would be that the second version is easier user-handleable than the first one, which would make the first one a reasonable choice as a default to me, and if it can be handled to act like the second version on the user site (e.g. handling the change between from and to bus or adding two dc lines with a lower limit of 0) that would leave out the parameter as well. 

3. p_loss1 is the constant power loss and p_loss2 the relatie one. So overall the loss is calculated as P_loss = p_loss1 + ploss_2 * p_mw.
Therefore if (for whatever reason, let it be a stupid edge case or user error) p_mw < p_loss1 i think the current implementation would lead to both generators (at the from AND and the to bus) to take power out the system. 

Let me look a bit more into timeseries calculations, I have not used them as of now, but eventually will anyway. 

To clarify: A timeseries test example?

**Comment by vogt31337** [comment]
@fbonhoff could you have a look on how I've implemented it currently? I've merged #2731 already.

'to the line' - I think in that case you really want to use a VSC and implement a multi terminal HVDC system. The dcline afaik is to be a simple point to point dc connection which allows a value to be set.

**Comment by KS-HTK** [comment]
@fbonhoff is there any update to this?
Maybe provide some clarity to what the reproducible example should output once the issue is fixed. Currently I get two differing values, my guess is they should be the same? So this issue is still relevant.

**Comment by fbonhoff** [comment]
Ah sorry, I had to roll back the Pandapower version I am using due to an error of loading .m files. 
I have some time to test, I will get to it during the week, check the implementation of @vogt31337 and also give an update to the parsing problem of .m files. 

What do you mean with you get 2 differing results? For which two evaluations?

**Comment by KS-HTK** [comment]
When running the reproducible example on the develop branch i get two values as an output.

```
643.2431452589881
----------------------------------------------------------------------
638.8613510465881
```

I have not looked to closely at what the code tries to do and evaluates. I was just looking over issues to figure out if they are still relevant. This issues mentioned a merged PR so I was unsure of its state.

**Comment by KS-HTK** [comment]
Is there any update to this? Did #2731 close this issue?
